### PR TITLE
issue02

### LINKED ISSUE #2: Save Image Link
#### As a user I want to save a link to an image.

__Story__:

Now that we have an awesome landing page for our users to experience it is time
to introduce our first feature! In general, we want to deliver value to our
customers as fast as possible. As a result, we have decided that we want our
users to be able to save links to images, and immediately see that we have
saved the URL to the image.

One could argue that simply saving the image, and not showing it, provides some
value. Yes, you could split this story here. However, because saving image URLs
and viewing those images are both relatively small additions that work well
together, it is acceptable to complete them together. Note, on the other hand,
that we do not yet want to implement the image index. Its addition adds too
much complexity, so we’ll save that for a future story.

In general, feature addition, or stories, should be split into the smallest
deliverable chunk that provides value to your users with the caveat that it
might make sense to combine two, sometimes more, tiny and related features into
a single story.

__Acceptance criteria__:
- [x] The landing page links to the image link submission form.
- [x] The link is entered through a form.
- [x] After the form is saved, the link is persisted in the database.
- [x] After the form is submitted, the user is redirected to a page displaying
  the image.
- [x] I cannot successfully save an image with an invalid URL.
- [x] An error message is associated with the appropriate input field on
  failure.
- [x] A flash message is displayed indicating the image was created successfully

__Discussion Topic__:
- [x] What is a valid image URL?

__Dependencies__:
- Deployment Environment

__Note__: Many image sharing sites will permit one to upload images, or
download images when provided a link. For this project we will only work with
the link, i.e., URL, to various images.

__Note__: To understand the database tables created by the `acts_as_taggable` gem, you may find it useful to use a program like the [DB Browser for SQLite](http://sqlitebrowser.org/).

### PR DESCRIPTION
**Pull request by lei-appfolio**
Closes #02

---
Issue #2 PR

The following is a checklist to remind you of the steps to take when opening a
new PR. The horizontal line above and everything below can be removed from the
pull request description.

- [x] I have replaced `X` above with the associated issue number.
- [x] I have changed the issue's label to "pull request".
- [x] I have assigned the designated reviewer to this PR.
- [x] I have set the appropriate label on this PR.
- [x] I have written clear commit messages ([ref](http://chris.beams.io/posts/git-commit/)).

If needed, please review the document on the
[Image Sharing Project](https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project)
for information on how labels should be used.

### CODE REVIEW

File: app/helpers/images_helper.rb [hunk 1]
```diff
@@ -0,0 +1,2 @@
+module ImagesHelper
```
Comment (bhargav265) [inline app/helpers/images_helper.rb]: I dont think you need this

File: app/models/image.rb [hunk 2]
```diff
@@ -0,0 +1,3 @@
+class Image < ApplicationRecord
```
Comment (bhargav265) [inline app/models/image.rb]: You can put this in a separate commit as `added Image model`

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -1,6 +1,4 @@
 class ApplicationController < ActionController::Base
   protect_from_forgery with: :exception
-  def home
-    @name = 'World'
-  end
+  def home; end
```
Comment (bhargav265) [inline app/controllers/application_controller.rb]: You can put this in a different commit 'Added Home page' after you commit the new page.

File: app/views/images/new.html.erb [hunk 4]
```diff
@@ -0,0 +1,8 @@
+<h1>New Image URL</h1>
+
```
Comment (bhargav265) [inline app/views/images/new.html.erb]: Put this in a separate commit

File: config/routes.rb [hunk 5]
```diff
@@ -1,4 +1,6 @@
 Rails.application.routes.draw do
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
   root 'application#home'
+
+  resources :images, only: %i[new create]
```
Comment (bhargav265) [inline config/routes.rb]: I think you can separate this out too

File: test/fixtures/images.yml [hunk 6]
```diff
@@ -0,0 +1,7 @@
+# Read about fixtures at http://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
```
Comment (bhargav265) [inline test/fixtures/images.yml]: I think you can remove this too

File: test/controllers/images_controller_test.rb [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+  end
+
+  def test_create__fail
+    assert_no_difference('Image.count') do
+      image_params = { web_url: 'Hello world' }
+      post images_path, params: { image: image_params }
+    end
+
+    assert_response :unprocessable_entity
+    assert_select 'span', 'is an invalid URL'
```
Comment (bhargav265) [inline test/controllers/images_controller_test.rb]: I think this is a good way of testing this `assert_equal "Some message", flash[:info]`. There might be other `span` in the future which would make this test fail

File: test/fixtures/images.yml [hunk 8]
```diff
@@ -0,0 +1,7 @@
...
+
+one:
+  web_url: MyText
```
Comment (bhargav265) [inline test/fixtures/images.yml]: You can remove these lines

Comment (bhargav265) [reply test/fixtures/images.yml]: Having the file means that for every test the table will be recreated

File: app/assets/javascripts/images.js [hunk 9]
```diff
@@ -0,0 +1,2 @@
+// Place all the behaviors and hooks related to the matching controller here.
```
Comment (bhargav265) [inline app/assets/javascripts/images.js]: Remove this file because we do not need it

File: app/assets/stylesheets/images.scss [hunk 10]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the images controller here.
```
Comment (bhargav265) [inline app/assets/stylesheets/images.scss]: Remove this file because we do not need it

File: config/routes.rb [hunk 11]
```diff
@@ -1,4 +1,5 @@
 Rails.application.routes.draw do
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
   root 'application#home'
+  resources :images, only: %i[new create]
```
Comment (bhargav265) [inline config/routes.rb]: You just need to add the `new` action here. You can add the `create` action in the commit where you made the create image functionality

File: test/controllers/images_controller_test.rb [hunk 12]
```diff
@@ -10,4 +10,23 @@ def test_new
     assert_select '.simple_form'
     assert_select '#new_image'
   end
+
+  def test_create__succeed
+    assert_difference('Image.count', 1) do
+      image_params = { web_url: 'https://imgur.com/gallery/Lz2m84C' }
+      post images_path, params: { image: image_params }
+    end
+
```
Comment (bhargav265) [inline test/controllers/images_controller_test.rb]: This needs a flash message according to the acceptance criteria

File: config/initializers/simple_form.rb [hunk 13]
```diff
@@ -0,0 +1,181 @@
+# frozen_string_literal: true
+
+# Uncomment this and change the path if necessary to include your own
+# components.
```
Comment (bhargav265) [inline config/initializers/simple_form.rb]: This should be in the commit where you added the simple form

Comment (lei-appfolio) [reply config/initializers/simple_form.rb]: [Add simple_form] and [Initialize simple_form] are two commits

File: config/routes.rb (hunk 11 above)
Comment (lei-appfolio) [reply config/routes.rb]: Here we must have create to make sure rails knows what is images_path

File: test/fixtures/images.yml (hunk 8 above)
Comment (lei-appfolio) [reply test/fixtures/images.yml]: OK

File: app/assets/javascripts/images.js (hunk 9 above)
Comment (lei-appfolio) [reply app/assets/javascripts/images.js]: OK

File: app/assets/stylesheets/images.scss (hunk 10 above)
Comment (lei-appfolio) [reply app/assets/stylesheets/images.scss]: OK

File: test/controllers/images_controller_test.rb (hunk 12 above)
Comment (bhargav265) [inline test/controllers/images_controller_test.rb]: Need to assert the `notice`

File: app/controllers/images_controller.rb [hunk 14]
```diff
@@ -1,2 +1,5 @@
 class ImagesController < ApplicationController
+  def new
+    @image = Image.new
```
Comment (bhargav265) [inline app/controllers/images_controller.rb]: Fix up this to the previous commit
